packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts

```diff
@@ -0,0 +1,118 @@
+const { toBN } = Web3.utils;
+
+import winston from "winston";
+
+import { createEtherscanLinkMarkdown, createFormatFunction, PublicNetworks } from "@uma/common";
+import { getAbi } from "@uma/contracts-node";
+import { InsuredBridgeL1Client, InsuredBridgeL2Client, GasEstimator } from "@uma/financial-templates-lib";
+
+import type { BN } from "@uma/common";
+
+export class CrossDomainFinalizer {
+  constructor(
+    readonly logger: winston.Logger,
+    readonly gasEstimator: GasEstimator,
+    readonly l1Client: InsuredBridgeL1Client,
+    readonly l2Client: InsuredBridgeL2Client,
+    readonly account: string,
+    readonly crossDomainFinalizationThreshold: number = 5
+  ) {}
+  async checkForBridgeableL2TokensAndBridge() {
+    this.logger.debug({ at: "AcrossRelayer#CrossDomainFinalizer", message: "Checking bridgeable L2 tokens" });
+
+    const whitelistedTokenEvents = await this.l1Client.bridgeAdmin.getPastEvents("WhitelistToken", { fromBlock: 0 });
+
+    // Extract the l2Tokens that have been whitelisted. The new Set syntax acts to remove any duplicates from the array.
+    // This would be the case if a l2 token was re - whitelisted after upgrading the bridgePool contract on L1.
+    const whitelistedL2Tokens = [...new Set(whitelistedTokenEvents.map((event) => event.returnValues.l2Token))];
+
+    // Check if any of the whitelisted l2Tokens are bridgeable. Do this in one parallel call. Returns an array of bool
+    // for each l2Token, describing if it can be bridged from L2->L1.
+    const canBridge = await Promise.all(
+      whitelistedL2Tokens.map((l2Token) => this.l2Client.bridgeDepositBox.methods.canBridge(l2Token).call())
+    );
+
+    // For each canBridge result, check if it is true. If so, then we can bridge the token.
+    const bridgeableL2Tokens = whitelistedL2Tokens.filter((_, index) => canBridge[index]);
+
+    // Finally, iterate over the bridgeable l2Tokens and bridge them.
+    if (bridgeableL2Tokens.length == 0) {
+      this.logger.debug({ at: "AcrossRelayer#CrossDomainFinalizer", message: "No bridgeable L2 tokens" });
+      return;
+    }
+    // Track the account nonce and manually increment on each TX. We need to do this because the L2 transactions
+    // process quicker than the infura node updates and we need to avoid the nonce collision.
+    let nonce = await this.l2Client.l2Web3.eth.getTransactionCount(this.account);
+    for (const l2Token of bridgeableL2Tokens) {
+      // For each bridgeable L2Token, check the balance in the deposit box. If it is greater than
+      // crossDomainFinalizationThreshold, as a percentage, then we can bridge it.
+
+      try {
+        const { symbol, decimals, l2PoolBalance } = await this._getL2TokenInfo(l2Token);
+        const l1PoolReserves = await this._getL1PoolReserves(l2Token);
+
+        if (l2PoolBalance.gt(toBN(this.crossDomainFinalizationThreshold).mul(l1PoolReserves).div(toBN(100))))
+          nonce = await this._bridgeL2Token(l2Token, nonce, symbol, decimals);
```

Comment:
It's kind of weird to return the `nonce`. I assume the `nonce` is getting incremented inside `_bridgeL2Token` which is potentially dangerous since it's not made explicit. I think it might be more clear to increment `nonce++` before calling `await this._bridgeL2Token()` and then passing in the new `nonce`.

This would make this logic more functional and easier to reason about